kolibri/plugins/management/assets/src/actions.js

```diff
@@ -69,110 +79,142 @@ function _managePageTitle(title) {
-    FacilityUserModel.attributes.roles[0].id : null;
-  const oldRole = FacilityUserModel.attributes.roles.length ?
-    FacilityUserModel.attributes.roles[0].kind : 'learner';
-
-  if (oldRole !== role) {
-  // the role changed
-    if (oldRole === 'learner') {
-    // role is admin or coach.
-      const rolePayload = {
-        user: id,
-        collection: FacilityUserModel.attributes.facility,
-        kind: role,
-      };
-      const RoleModel = RoleResource.createModel(rolePayload);
-      RoleModel.save(rolePayload).then((newRole) => {
-        FacilityUserModel.save(payload).then(responses => {
-          // force role change because if the role is the only changing attribute
-          // FacilityUserModel.save() will not send request to server.
-          responses.roles = [newRole];
-          store.dispatch('UPDATE_USERS', [responses]);
-        })
-        .catch(error => { coreActions.handleApiError(store, error); });
-      });
-    } else if (role !== 'learner') {
-    // oldRole is admin and role is coach or oldRole is coach and role is admin.
-      const OldRoleModel = RoleResource.getModel(oldRoldID);
-      OldRoleModel.delete().then(() => {
-      // create new role when old role is successfully deleted.
-        const rolePayload = {
-          user: id,
-          collection: FacilityUserModel.attributes.facility,
-          kind: role,
-        };
-        const RoleModel = RoleResource.createModel(rolePayload);
-        RoleModel.save(rolePayload).then((newRole) => {
-        // update the facilityUser when new role is successfully created.
-          FacilityUserModel.save(payload).then(responses => {
-            // force role change because if the role is the only changing attribute
-            // FacilityUserModel.save() will not send request to server.
-            responses.roles = [newRole];
-            store.dispatch('UPDATE_USERS', [responses]);
-          })
-          .catch(error => { coreActions.handleApiError(store, error); });
-        });
-      })
-      .catch(error => { coreActions.handleApiError(store, error); });
-    } else {
-    // role is learner and oldRole is admin or coach.
-      const OldRoleModel = RoleResource.getModel(oldRoldID);
-      OldRoleModel.delete().then(() => {
-        FacilityUserModel.save(payload).then(responses => {
-          // force role change because if the role is the only changing attribute
-          // FacilityUserModel.save() will not send request to server.
-          responses.roles = [];
-          store.dispatch('UPDATE_USERS', [responses]);
-        })
-        .catch(error => { coreActions.handleApiError(store, error); });
+function updateUser(store, stateUser) {
+  //payload needs username, fullname, and facility
+  const userID = stateUser.id;
+  const savedUserModel = FacilityUserResource.getModel(userID);
+  let savedUser = savedUserModel.attributes;
+  let roleAssigned = Promise.resolve(savedUserModel.attributes);
+  let changedValues = {};
```

Comment:
Hm. Alright, I think I see what you mean. I can see how it's easier to read. I can't say I'm used to this sort of functional programming (school's focus is Object Oriented stuff) and "pure" function writing, but I think I understand how separating these concerns can make things easier to debug in the long run.

I'll give it a shot and resubmit the PR. Thanks for looking this over in such detail.